Maintain per-thread timer_id rather than create/delete frequently

jiebinn opened the pull request:
The QueryProfiler will frequently create/delete timer_id globally, which will result in heavy kernel lock contention.
The idea is to maintain thread-local timer_id. Before create the timer_id, it should check whether there is a timer_id already. And we could stop the timer by timer_settime() rather than delete the timer_id with timer_delete().

Apply the patch and run clickbench on latest 65d671b7c72c ClickHouse with SPR 112 x 2 vCPUs. Query 4,0,5,3,15,32 have 17.5%,14.4%,8.3%,7.9%, 7.1%,5.8% performance gain. The overall geomean has got 2.5% performance gain.

According to the #49965 , the patch has got 18% overall geomean performance in SSB benchmark in recent ClickHouse upstream with ICX 80 x 2 vCPUs.

<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/run_check.py
-->
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Improve performance with enabled QueryProfiler using thread-local timer_id instead of global object.

jiebinn committed:
Maintain per-thread timer_id rather than create/delete frequently

The QueryProfiler will frequently create/delete timer_id globally, which
will result in heavy kernel lock contention.
The idea is to maintain thread-local timer_id. Before create the
timer_id, it should check whether there is a timer_id already. And we
could stop the timer by timer_settime() rather than delete the timer_id
with timer_delete().

Apply the patch and run clickbench on latest 65d671b7c72c ClickHouse with
SPR 112 x 2 vCPUs. Query 4, 0, 5, 3, 15, 32 have 17.5%, 14.4%, 8.3%, 7.9%,
7.1%, 5.8% performance gain. The overall geomean has got 2.5%
performance gain.

Signed-off-by: Jiebin Sun <jiebin.sun@intel.com>

jiebinn commented:
@davenger Hi Alexander, what else should I do for the patch? Thank you.

jiebinn commented:
> I think it would be nice if the timer was deleted when the thread terminates. Maybe you can wrap it in a RAII class?

Hi aiven, Thanks for your comment. Sorry for the late reply. The timer would be deleted once the thread pool is terminated. We do not need to delete the timer when it is detached as the timer is always existed in each query.

nickitat commented:
@jiebinn I agree with aiven-sal@ that we're better to wrap it to make sure that timer (corresponding kernel object) will be freed on thread termination.

robot-ch-test-poll1 commented:
<!-- automatic status comment for PR #48778 from jiebinn/ClickHouse:QueryProfiler-timerID -->
This is an automated comment for commit 35cb9293eeba5d0f06afc1417a8532fd10cb32ee with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/48778/35cb9293eeba5d0f06afc1417a8532fd10cb32ee/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

jiebinn commented:
Hi @aiven-sal @nickitat, thanks for your advice. I have reviewed the related code and completed the v2 patch since last week. Run the code style check script and the performance test before submitting the PR. The performance gain is the same with the previous version.

v2: It has packed the Timer and related methods into the class. The timer_id has been deleted when the thread terminates.

jiebinn commented on src/Common/QueryProfiler.cpp:
```diff
@@ -185,15 +225,7 @@ template <typename ProfilerImpl>
 void QueryProfilerBase<ProfilerImpl>::tryCleanup()
 {
 #if USE_UNWIND
-    if (timer_id.has_value())
```
Done.

jiebinn commented on src/Common/QueryProfiler.h:
```diff
@@ -27,6 +27,22 @@ namespace DB
+class Timer
+{
+public:
+    Timer(){ timer_id = std::optional<timer_t>(std::nullopt); }
```
Done

jiebinn commented on src/Common/QueryProfiler.cpp:
```diff
@@ -82,9 +82,88 @@ namespace ErrorCodes
+        struct itimerspec timer_spec = {.it_interval = stop_timer, .it_value = stop_timer};
+        int err = timer_settime(*timer_id, 0, &timer_spec, nullptr);
+        if (err)
+            throw Exception(ErrorCodes::CANNOT_SET_TIMER_PERIOD, "Failed to stop query profiler timer");
```
Done.

jiebinn commented on src/Common/QueryProfiler.cpp:
```diff
@@ -82,9 +82,88 @@ namespace ErrorCodes
+
+Timer::~Timer()
+{
+    tryCleanup();
```
Done.

jiebinn replied:
Use 'Timer::~Timer() noexcept(false)' instead.

jiebinn replied:
@aiven-sal Got it. Thanks for the details and reference. So, we should keep the same with the original code and LOG_* functions and chassert instead of throw. I have updated the new branch code.

jiebinn commented on src/Common/QueryProfiler.cpp:
```diff
@@ -82,9 +82,88 @@ namespace ErrorCodes
+    if (timer_id.has_value())
+    {
+        int err = timer_delete(*timer_id);
+        throw Exception(ErrorCodes::CANNOT_DELETE_TIMER, "Failed to delete query profiler timer");
```
Done.

jiebinn commented:
Hi @aiven-sal , I have pushed another commit to fix the PR issue. Please help to review and confirm if we have got the same point.

jiebinn committed:
Pack the timer and delete the timer_id when thread terminates

Pack the timer and related methods into the class. Delete the timer_id
when the thread terminates.

According to the issue (ClickHouse#49965),
all of the SSB queries benefit from this optimization, some have even got
improved by ~30% and the overall QPS could be significantly improved by ~18%.

Signed-off-by: Jiebin Sun <jiebin.sun@intel.com>

jiebinn committed:
Merge remote-tracking branch 'upstream/master' into QueryProfiler-timerID

jiebinn committed:
Fix the review issue of QueryProfiler Timer from PR
https://github.com/ClickHouse/ClickHouse/pull/48778.

Signed-off-by: Jiebin Sun <jiebin.sun@intel.com>

jiebinn commented on src/Common/QueryProfiler.h:
```diff
@@ -27,6 +27,22 @@ namespace DB
+{
+public:
+    Timer(){ timer_id = std::optional<timer_t>(std::nullopt); }
+    ~Timer();
```
I have missed that.

alexey-milovidov commented on src/Common/QueryProfiler.cpp:
```diff
@@ -85,6 +85,83 @@ namespace ErrorCodes
+    {
+        int err = timer_delete(*timer_id);
+        if (err)
+            LOG_ERROR(log, "Failed to delete query profiler timer {}", errnoToString());
```
LOG_ERROR can throw, let's wrap it to catch in destructor.

jiebinn replied:
Thanks for the review. Catch the throw error of LOG_ERROR in the destructor both in Timer and QueryProfilerBase.

jiebinn commented:
Hi @nickitat, thanks for review the patch. What else should I do for this PR or any suggestion for it?

jiebinn commented on src/Common/QueryProfiler.cpp:
```diff
@@ -85,6 +85,87 @@ namespace ErrorCodes
+        timer_t local_timer_id;
+        if (timer_create(clock_type, &sev, &local_timer_id))
+            /// In Google Cloud Run, the function "timer_create" is implemented incorrectly as of 2020-01-25.
+            /// https://mybranch.dev/posts/clickhouse-on-cloud-run/
```
Delete the legacy comment.

alexey-milovidov replied:
Keep it.

jiebinn replied:
Keep the original code here.

azat commented on src/Common/QueryProfiler.cpp:
```diff
@@ -85,6 +85,92 @@ namespace ErrorCodes
+        struct timespec stop_timer{.tv_sec = 0, .tv_nsec = 0};
+        struct itimerspec timer_spec = {.it_interval = stop_timer, .it_value = stop_timer};
+        int err = timer_settime(*timer_id, 0, &timer_spec, nullptr);
+        if (err)
+            LOG_ERROR(log, "Failed to stop query profiler timer {}", errnoToString());
+        chassert(!err && "Failed to stop query profiler timer");
```
```suggestion
        if (err)
            throwFromErrno("Failed to stop query profiler timer", ErrorCodes::LOGICAL_ERROR);
```

You may think about `LOGICAL_ERROR` as about `chassert`

azat commented on src/Common/QueryProfiler.cpp:
```diff
@@ -85,6 +85,92 @@ namespace ErrorCodes
 }
 
+#if USE_UNWIND
+Timer::Timer():log(&Poco::Logger::get("Timer")){}
```
```suggestion
Timer::Timer()
    : log(&Poco::Logger::get("Timer"))
{}
```

azat commented on src/Common/QueryProfiler.cpp:
```diff
@@ -85,6 +85,92 @@ namespace ErrorCodes
+
+Timer::~Timer()
+{
+    try {
+        cleanup();
+    } catch(...) {}
```
Hiding exceptions is not a good idea, consider using something like this:

```suggestion
    try
    {
        cleanup();
    }
    catch(...)
    {
        tryLogCurrentException(log);
    }
```

azat reviewed (commented):
My main suggestion is to add two separate `CurrentMetrics` for created and active timers, this will increase introspection.

P.S. Sorry for being so late.

jiebinn committed:
Update src/Common/QueryProfiler.cpp

Co-authored-by: Azat Khuzhin <a3at.mail@gmail.com>

jiebinn committed:
Update src/Common/QueryProfiler.cpp

Co-authored-by: Azat Khuzhin <a3at.mail@gmail.com>

jiebinn committed:
Update src/Common/QueryProfiler.cpp

Co-authored-by: Azat Khuzhin <a3at.mail@gmail.com>

jiebinn commented:
> My main suggestion is to add two separate `CurrentMetrics` for created and active timers, this will increase introspection.
> 
> P.S. Sorry for being so late.

Hi @azat , I have added two metrics for created and active timers in latest patch. Please help to review. Thanks.

jiebinn committed:
Add two separate CurrentMetrics for created and active timers
in QueryProfiler.

Signed-off-by: Jiebin Sun <jiebin.sun@intel.com>

jiebinn committed:
Merge branch 'master' into QueryProfiler-timerID

jiebinn committed:
Merge branch 'master' into QueryProfiler-timerID

jiebinn commented:
@nickitat 
The CI check tells that
'Jun 09 10:30:07 /build/src/Disks/ObjectStorages/AzureBlobStorage/AzureObjectStorage.h:11:10: fatal error: 'azure/storage/blobs.hpp' file not found
Jun 09 10:30:07 #include <azure/storage/blobs.hpp>'
It seems not related to the patch.

nickitat commented:
> @nickitat
> The CI check tells that
> 'Jun 09 10:30:07 /build/src/Disks/ObjectStorages/AzureBlobStorage/AzureObjectStorage.h:11:10: fatal error: 'azure/storage/blobs.hpp' file not found
> Jun 09 10:30:07 #include <azure/storage/blobs.hpp>'
> It seems not related to the patch.

we had a bug unfortunately #50770

nickitat committed:
Merge branch 'master' into QueryProfiler-timerID

jiebinn committed:
Merge branch 'master' into QueryProfiler-timerID
